### PR TITLE
Two way streams fixes - take 2

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Re-add #19763 - fixes are individual commits.

Fixes races observed with new tests and tweak tests for flaky operations.

Also fixes a goroutine leak with (unused) two-way streams.

## How to test this PR?

In internal/grid, use `go test -v --count=100 -race -timeout=30m -failfast` or similar.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Unit tests added/updated

### CODE REVIEW

**Review by harshavardhana — APPROVED** [review]

**Comment by klauspost** [comment]
@krisis @vadmeste PTAL. Now (just post release) would be a good time to get this merged.

File: internal/grid/connection.go [hunk 1]
```diff
@@ -1406,6 +1412,10 @@ func (c *Connection) handleRequest(ctx context.Context, m message, subID *subHan
 }
 
 func (c *Connection) handlePong(ctx context.Context, m message) {
+	if m.MuxID == 0 && m.Payload == nil {
+		atomic.StoreInt64(&c.LastPong, time.Now().Unix())
```
Comment (harshavardhana) [inline internal/grid/connection.go]:
```suggestion
		atomic.StoreInt64(&c.LastPong, time.Now().UnixNano())
```

?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: internal/grid/connection.go (hunk 1 above)
Comment (klauspost) [reply internal/grid/connection.go]: The one on `Connection` is secs. The one on `muxClient` is nano. Updated it so I could do the tests without them taking minutes.

Comment (klauspost) [reply internal/grid/connection.go]: Maybe I should switch Connection to just nano for consistency. It will not overflow for a long time anyway.

Comment (klauspost) [reply internal/grid/connection.go]: Friday, April 11, 2262 will be a bad day :D

**Review by krisis — APPROVED** [review]
LGTM. Out of curiosity, where would two way streams be used in MinIO's internode communications?

**Comment by klauspost** [comment]
@krisis Locking is most obvious. That way the request will track the lifetime of the lock request to the remote.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**